src/webui/api/torrentscontroller.cpp

```diff
@@ -250,6 +257,89 @@ namespace
             idList << BitTorrent::TorrentID::fromString(hash);
         return idList;
     }
+
+    QJsonObject serializeTorrentInfo(const BitTorrent::TorrentInfo &info, const QList<BitTorrent::TrackerEntry> &trackers)
+    {
+        QJsonArray trackersArr;
+        for (const auto &tracker : trackers)
+        {
+            trackersArr << QJsonObject
+            {
+                {KEY_TRACKER_URL, tracker.url},
+                {KEY_TRACKER_TIER, tracker.tier}
+            };
+        }
+
+        QJsonArray files;
+        for (int fileIndex = 0; fileIndex < info.filesCount(); ++fileIndex)
+        {
+            files << QJsonObject
+            {
+                {KEY_FILE_INDEX, fileIndex},
+                {KEY_FILE_NAME, info.filePath(fileIndex).toString()},
+                {KEY_FILE_SIZE, info.fileSize(fileIndex)}
+            };
+        }
+
+        QJsonArray webseeds;
+        for (const QUrl &webseed : info.urlSeeds())
+        {
+            webseeds << QJsonObject
+            {
+                {KEY_WEBSEED_URL, webseed.toString()}
+            };
+        }
+
+        return QJsonObject {
+            {KEY_TORRENT_INFOHASHV1, info.infoHash().v1().toString()},
+            {KEY_TORRENT_INFOHASHV2, info.infoHash().v2().toString()},
+            {KEY_TORRENT_NAME, info.name()},
+            {KEY_TORRENT_ID, info.infoHash().toTorrentID().toString()},
+            {KEY_TORRENTINFO_TRACKERS, trackersArr},
+            {KEY_PROP_TOTAL_SIZE, info.totalSize()},
+            {KEY_PROP_PIECES_NUM, info.piecesCount()},
+            {KEY_PROP_PIECE_SIZE, info.pieceLength()},
+            {KEY_PROP_CREATED_BY, info.creator()},
+            {KEY_PROP_PRIVATE, info.isPrivate()},
+            {KEY_PROP_CREATION_DATE, Utils::DateTime::toSecsSinceEpoch(info.creationDate())},
+            {KEY_PROP_COMMENT, info.comment()},
+            {KEY_TORRENTINFO_FILES, files},
+            {KEY_TORRENTINFO_WEBSEEDS, webseeds}
+        };
+    }
+
+    QJsonObject serializeTorrentInfo(const BitTorrent::TorrentDescriptor &torrentDescr)
+    {
+        const BitTorrent::TorrentInfo &info = torrentDescr.info().value();
+
+        QList<BitTorrent::TrackerEntry> trackers = torrentDescr.trackers();
+        // these fields are inconsistent - sometimes trackers are stored in TorrentInfo, sometimes in TorrentDescriptor
```

Comment:
It's not clear to me how the situation in the linked code will now be handled. In cases where the trackers were stored in `TorrentInfo`, will they now be inaccessible? Surely there's some way to access them but I haven't been able to identify it.

Reply:
> In cases where the trackers were stored in `TorrentInfo`, will they now be inaccessible?

Have you really encountered a situation where a TorrentDescriptor created from a torrent file does not return trackers through its own method, but they are contained in the TorrentInfo? This could mean some kind of bug in TorrentDescriptor implementation that needs to be fixed.

Reply:
Yes, I ran into this during my testing. I'll see if I can reproduce again and identify the exact circumstances.

Reply:
Also relevant: https://github.com/qbittorrent/qBittorrent/pull/21015#discussion_r1660519462

Reply:
> Also relevant: [#21015 (comment)](https://github.com/qbittorrent/qBittorrent/pull/21015#discussion_r1660519462)

This is only a consequence of what was said above:
> `TorrentInfo` is no more supposed to provide anything except info-section related metadata.

We must proceed from the fact that any properties of `TorrentInfo` that are not related to the info-section will contain invalid data and should not be used currently. They were forgotten to be removed from `TorrentInfo` class at one time. #21084 should eliminate this discrepancy.

Reply:
I'll switch away from using those `TorrentInfo` fields in preparation.

Reply:
> Yes, I ran into this during my testing. I'll see if I can reproduce again and identify the exact circumstances.

I identified the problem and it's in my code rather than an existing issue. In `onMetadataDownloaded` I was setting the torrent info on an otherwise empty `TorrentDescriptor`. The solution was to persist the initial `TorrentDescriptor` from parsing the magnet URI. I'll push up a fix.